# Permutations and Combinations

In this lesson, we'll direct our focus not so much toward calculating probabilities, but toward calculating the number of outcomes associated with various random experiments. We'll learn some powerful counting techniques that will allow us to answer questions like:

- What is the probability of cracking a 4-digit PIN code using the code 8362?

- What is the probability of cracking a 6-digit PIN code using the code 348821?

- What is the probability of winning the big prize in a state lottery if we use the numbers (3, 20, 37, 44, 45, 49)?


We begin with considering a composite experiment A1A2 made of two different experiments, which we denote by "A1" and "A2":

- A1: flipping a coin

- A2: throwing a six-sided die

There are 12 possible outcomes associated with A1A2:

![probability-pic-5](https://raw.githubusercontent.com/tongNJ/Dataquest-Online-Courses-2022/main/Pictures/probability-pic-5.PNG)

**Number of outcomes  = a ⋅ b = 2 ⋅ 6 = 12**



### Test - Let's say we're interested in finding the probability of cracking a 4-digit PIN code using the code 8362 (we chose 8362 randomly).

#### This is the example of sample WITH REPLACEMENT!

- Total number of outcomes of 4-digit PIN Code = 10 * 10 * 10 * 10 = 10000

- The probability of cracking 4-digit Pin with 8362 = 1/10000 = 0.0001

In this case, each PIN code represents a certain arrangement where the order of the individual digits matters. Because order matters, the code 1289 is different than the code 9821, even though both are composed of the same four digits: 1, 2, 8 and 9. If the order of digits didn't matter, 1289 would be the same as 9821.


More generally, a certain arrangement where the order of the individual elements matters is called a **permutation**. For instance, there are 10,000 possible permutations for a 4-digit PIN code (in other words, there are 10,000 digit arrangements where the order of the digits matters).



---------------------------------------------------------------------------------------------------------------------

Let's now turn our attention to another example, where we're interested in finding the total number of possible permutations for the numbers {5, 6, 7} — one possible permutation would be 567, another 657, and so on.

However, this time we want to form the permutations by sampling ***without replacement***. For instance, if we sample once without replacement from the set {5, 6, 7} and get a 5, we don't put the 5 back in the set, and only {6, 7} remains for the second round of sampling.

To form a permutation like 567 or 657, we need to sample three times from the set {5, 6, 7}. Since we sample three times, we can break down the entire process into three experiments:

- E1, which has three possible outcomes: {5, 6, 7}

- E2, which has two possible outcomes left (because we sample without replacement we leave out the element we got at E1; if we get a 6 at E1, only {5, 7} is left for E2)

- E3, which has one possible outcome left

Using the extended rule of product, we see we have a total of six outcomes for the composite experiment E1E2E3:

**Number of outcomes  = n! = 3! = 3 ⋅ 2 ⋅ 1 = 6**

### REMEMBER!  SAMPLE WITHOUT REPLACEMENT!!


![probability-pic-6](https://raw.githubusercontent.com/tongNJ/Dataquest-Online-Courses-2022/main/Pictures/probability-pic-6.PNG)


To summarize, when we sample without replacement, the number of permutations is given by the formula:

**Permutation = n!**

In [14]:
# the function I wrote
def factorial(x):
    y = 1
    while x > 1:
        y = y * x * (x-1)
        x = x-2
    return y

#########################################################

# the function from answer sheet
def factorial(n):
    final_product = 1
    for i in range(n,0,-1):
        final_product *= i
    return final_product

In [15]:
factorial(5)

120

### Texas hold'em

Taxas hold'em, which is a variation of poker, players are interested in having a winning 5-card poker hand. To find the total number of possible permutations for a 5-card poker hand, we can start by considering we're sampling without replacement five times from a standard 52-card deck:


E1 = 52 outcomes
E2 = 51 outcomes
E3 = 50 outcomes
E4 = 49 outcomes
E5 = 48 outcomes

So total possible permutatins = 52 * 51 * 50 * 49 * 48 = 311,875,200 outcomes!!

More generally, when we have a group of n objects, but we're taking only k objects, the number of permutations (which we abbreviate as "P") is:

**$_{n}P_{k}$ = n ⋅ (n−1) ⋅ (n−2) ⋅ … ⋅ (n−k+1)**

### $_{n}P_{k}$ = n!  /  (n - k)!

In [20]:
def permutation(n,k):
    n_P_k = factorial(n) / factorial(n-k)
    return n_P_k

In [21]:
permutation(52,5)

311875200.0

#### Test
A password manager software generates 16-character passwords from a list of 127 characters (the list contains numbers, letters, or other symbols). Assume the sampling from the list is done randomly and without replacement, and find the probability of cracking a password generated by this software if we're using the password "@*AmgJ(UL3Yl726x", which has 16 characters. Assign your answer to p_crack_pass.

In [22]:
p_crack_pass = 1/permutation(n=127,k=16)
print(p_crack_pass)

5.851813813338265e-34


### Cont. Texas hold'em

Previously, we mentioned players in Texas hold 'em are interested in having a winning 5-card poker hand. To find the number of permutations for a 5-card poker hand, we learned to use this formula:

However, remember that a permutation is an arrangement of elements where order matters. The three 5-card hands we see below, for instance, have the same cards and yet are considered different permutations (arrangements) because order matters.

![probability-pic-7](https://raw.githubusercontent.com/tongNJ/Dataquest-Online-Courses-2022/main/Pictures/probability-pic-7.PNG)

In a poker game, however, the order of cards is not important, so the three hands above would be considered identical. Consequently, we might be interested instead in ignoring the order and finding the number of unique card arrangements.

More generally, if the order of the elements in an arrangement doesn't matter, the arrangement is called a **combination**. Combinations are the opposite of permutations; the order of the elements does not matter for combinations, whereas the order does matter for permutations.

To find the number of unique card arrangements, we begin with the observation that each unique 5-card arrangement has 5! = 120 permutations. In other words, the five cards of a unique hand can be ordered in 120 different ways. In the diagram above, we see just three out the 120 permutations of that unique hand.

If each unique hand can be arranged in 5! = 120 ways and there are C total unique hands, then C ⋅ 5! gives us the total number of permutations of a 5-card hand:

**C ⋅ 5! = $_{52}P_{5}$**

**C = $_{52}P_{5}$  /  5!**

In [23]:
unique_combination = permutation(52,5)/factorial(5)

2598960.0

For a state lottery, six numbers are drawn randomly and without replacement from a set containing numbers from 1 to 49. Using the factorial() and the permutation() functions, find the total number of unique 6-number arrangements that could result. Assign your answer to c_lottery.

In [26]:
c = permutation(49,6) / factorial(6)
print(f'the state lottery has {c} number of unique combinations, and the prability of winner with one set of number is {round(100/c,10)}%')

the state lottery has 13983816.0 number of unique combinations, and the prability of winner with one set of number is 7.1511e-06%
